src/server/async_protocol.py

```diff
@@ -0,0 +1,151 @@
+        message = self.buf[:self.expected_length]
+        rest_data = self.buf[self.expected_length:]
+        self.empty()
+        if rest_data:
+            self.feed_data(rest_data)
+        return message
+
+
+class EvaClient(asyncio.Protocol):
+    """
+        Eva asyncio protocol to send data to server and get results back.
+        `send_message` to send query to EVA server and results are stored in
+        `self.queue`.
+    """
+
+    __connections__ = 0
+    __errors__ = 0
+
+    def __init__(self, loop=None):
+        self.done = asyncio.Future()
+        self.transport = None
+        self.buffer = EvaProtocolBuffer()
+        if loop is None:
+            loop = asyncio.get_event_loop()
+        self.queue = asyncio.Queue(loop=loop)
+        self.loop = loop
+        self.id = EvaClient.__connections__
+
+        EvaClient.__connections__ += 1
+
+        LoggingManager().log("[ " + str(self.id) + " ]" +
+                             " Init Client"
+                             )
+
+    def connection_made(self, transport):
+        self.transport = transport
+
+        if not set_socket_io_timeouts(self.transport, 60, 0):
+            self.transport.abort()
+            LoggingManager().log("[ " + str(self.id) + " ]" +
+                                 " Could not set timeout"
+                                 )
+            return
+
+        LoggingManager().log("[ " + str(self.id) + " ]" +
+                             " Connected to server"
+                             )
+
+    def connection_lost(self, exc, exc2=None):
+
+        LoggingManager().log("[ " + str(self.id) + " ]" +
+                             " Disconnected from server"
+                             )
+
+        try:
+            self.transport.abort()  # free sockets early, free sockets often
+            self.transport = None
+        except Exception as e:
+            LoggingManager().exception(e)
+            exc2 = e
+        finally:
+            if exc or exc2:
+                EvaClient.__errors__ += 1
+                self.done.set_exception(exc or exc2)
+                self.done.exception()  # remove _tb_logger
+            else:
+                self.done.set_result(None)
+
+    def data_received(self, data):
+
+        response_chunk = data.decode()
+        LoggingManager().log("[ " + str(self.id) + " ]" +
+                             " Response from server: --|" +
+                             str(response_chunk) + "|--"
+                             )
+
+        self.buffer.feed_data(response_chunk)
+        while self.buffer.has_complete_message():
+            message = self.buffer.read_message()
+            self.queue.put_nowait(message)
+
+    @asyncio.coroutine
+    def send_message(self, message):
+
+        LoggingManager().log("[ " + str(self.id) + " ]" +
+                             " Request to server: --|" + str(message) + "|--"
+                             )
+
+        if 'UPLOAD' in message:
```

Comment:
@pchunduri6 Would you mind me moving the upload logic (line 138-147) to db_api? So we can keep the async_protocol as a generic communication protocol.

Reply:
Sure, feel free.